code/modules/projectiles/projectile.dm

```diff
@@ -232,74 +231,61 @@
-				if(isnull(Angle))
-					Angle=round(Get_Angle(src,current))
-				if(spread)
-					Angle += (rand() - 0.5) * spread
-				var/matrix/M = new
-				M.Turn(Angle)
-				transform = M
-
-				var/Pixel_x=round(sin(Angle)+16*sin(Angle)*2)
-				var/Pixel_y=round(cos(Angle)+16*cos(Angle)*2)
-				var/pixel_x_offset = pixel_x + Pixel_x
-				var/pixel_y_offset = pixel_y + Pixel_y
-				var/new_x = x
-				var/new_y = y
-
-				while(pixel_x_offset > 16)
-					pixel_x_offset -= 32
-					pixel_x -= 32
-					new_x++// x++
-				while(pixel_x_offset < -16)
-					pixel_x_offset += 32
-					pixel_x += 32
-					new_x--
-
-				while(pixel_y_offset > 16)
-					pixel_y_offset -= 32
-					pixel_y -= 32
-					new_y++
-				while(pixel_y_offset < -16)
-					pixel_y_offset += 32
-					pixel_y += 32
-					new_y--
-
-				speed = round(speed)
-				step_towards(src, locate(new_x, new_y, z))
-				if(speed <= 1)
-					pixel_x = pixel_x_offset
-					pixel_y = pixel_y_offset
-				else
-					animate(src, pixel_x = pixel_x_offset, pixel_y = pixel_y_offset, time = max(1, (speed <= 3 ? speed - 1 : speed)))
-
-				if(original && (original.layer>=2.75) || ismob(original))
-					if(loc == get_turf(original))
-						if(!(original in permutated))
-							Bump(original, 1)
-				Range()
-			sleep(max(1, speed))
-	else //old projectile system
-		set waitfor = 0
-		while(loc)
-			if(!paused)
-				if((!( current ) || loc == current))
-					current = locate(Clamp(x+xo,1,world.maxx),Clamp(y+yo,1,world.maxy),z)
-				step_towards(src, current)
-				if(original && (original.layer>=2.75) || ismob(original))
-					if(loc == get_turf(original))
-						if(!(original in permutated))
-							Bump(original, 1)
-				Range()
-			sleep(1)
+
+	while(!QDELETED(src))
+		if(!paused)
+			if((!current || loc == current))
+				current = locate(Clamp(x + xo, 1, world.maxx), Clamp(y + yo, 1, world.maxy), z)
+			if(isnull(Angle))
+				Angle = round(Get_Angle(src, current))
+			if(spread)
+				Angle += (rand() - 0.5) * spread
+			var/matrix/M = new
+			M.Turn(Angle)
+			transform = M
+
+			var/Pixel_x = round(sin(Angle) + 16 * sin(Angle) * 2)
+			var/Pixel_y = round(cos(Angle) + 16 * cos(Angle) * 2)
+			var/pixel_x_offset = pixel_x + Pixel_x
+			var/pixel_y_offset = pixel_y + Pixel_y
+			var/new_x = x
+			var/new_y = y
+
+			while(pixel_x_offset > 16)
+				pixel_x_offset -= 32
+				pixel_x -= 32
+				new_x++ // x++
+			while(pixel_x_offset < -16)
+				pixel_x_offset += 32
+				pixel_x += 32
+				new_x--
+
+			while(pixel_y_offset > 16)
+				pixel_y_offset -= 32
+				pixel_y -= 32
+				new_y++
+			while(pixel_y_offset < -16)
+				pixel_y_offset += 32
+				pixel_y += 32
+				new_y--
+
+			speed = round(speed)
+			step_towards(src, locate(new_x, new_y, z))
+			if(speed <= 1)
+				pixel_x = pixel_x_offset
+				pixel_y = pixel_y_offset
+			else
+				animate(src, pixel_x = pixel_x_offset, pixel_y = pixel_y_offset, time = max(1, (speed <= 3 ? speed - 1 : speed)))
+
+			if(ismob(original) || (original && original.layer >= 2.75))
```

Comment:
was `if(original && (original.layer>=2.75) || ismob(original))`
changed it since mobs usually fire the bullets so it's a minor optimization. Also made the cases more clear since the parentheses didn't make sense

Reply:
This is actually more expensive. 2.75 is a layer that most mobs are at, therefore this will return `TRUE` and you won't be running the `istype` check with mobs.

Checking for a holder and its var value is going to short circuit wayyy more often than actually checking for `ismob`.

Also, this isn't the one doing the firing, this is the original target *clicked*.

Reply:
Ah that makes more sense. Will change it back to that order then